File: backend/src/main/java/com/example/backend/model/NotificationLog.java
package com.example.backend.model;

import com.example.backend.common.NotificationStatus;
import io.hypersistence.utils.hibernate.type.json.JsonBinaryType;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.Type;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "notification_logs")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class NotificationLog {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "log_id")
    private UUID id;

    @Column(name = "recipient", nullable = false)
    private String recipient;

    // N:1 Relationship with NotificationTemplate
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "template_id", nullable = false)
    private NotificationTemplate template;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 50)
    private NotificationStatus status;

    @Column(name = "source_event", length = 100)
    private String sourceEvent;

    @Type(JsonBinaryType.class)
    @Column(columnDefinition = "jsonb")
    private Map<String, Object> metadata;

    @Column(name = "sent_at")
    private LocalDateTime sentAt;

    // N:1 Relationship with User
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = true)
    private User user;

    @Builder.Default
    @Column(name = "retry_count")
    private Integer retryCount = 0;

    @Column(name = "error_message", columnDefinition = "TEXT")
    private String errorMessage;

    @Builder.Default
    @Column(name = "is_active")
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: frontend/src/types/notificationLog.ts
import type { BaseEntity, BaseFilterRequest } from "./common";
import type { NotificationStatus } from "./enum";

export interface NotificationLogResponse extends BaseEntity {
  recipient: string;
  templateId: string;
  templateName: string;
  status: NotificationStatus;
  sourceEvent: string;
  metadata: Record<string, unknown>; // JSON placeholder
  sentAt: string;
  userId: string;
  bookingId: string;
  retryCount: number;
  errorMessage: string;
}

export interface NotificationLogFilterRequest extends BaseFilterRequest {
  recipient?: string;
  templateId?: string;
  status?: NotificationStatus;
  sourceEvent?: string;
  retryCount?: number;
  sentFrom?: string;
  sentTo?: string;
  keyword?: string;
}
